types/jest/index.d.ts

```diff
@@ -340,7 +331,7 @@ declare namespace jest {
         fail(error?: string | { message: string }): any;
     }
 
-    type ProvidesCallback = (cb: DoneCallback) => any;
+    type ProvidesCallback = ((cb: DoneCallback) => void | undefined) | (() => Promise<unknown>);
```

Comment:
While this is correct for test suite & case functions, it's technically not for hooks because you can return any value so long as you don't have the `done` callback, as it's meant to make it easier compatibility with short-handed arrow functions.

I think we should make a new `HookCallback` to cover this (I'm impartial to if we create a new `Lifecycle`-like type or make `Lifecycle` take a generic of `ProvidesCallback | HookCallback`)

Reply:
Thanks for the review! :raised_hands:

> currently this type will require folks to return a promise even if they're not doing anything async

The `ProvidesCallback` type should allow them to return void or undefined or a promise if they don't use done (the tests verify this I believe). If they use done, they must return void or undefined. We could make this more explicit perhaps by changing this line to:

```ts
type ProvidesCallback = ((cb: DoneCallback) => void | undefined) | (() => Promise<unknown> | void | undefined);
```

---

> make it easier compatibility with short-handed arrow functions

I'm not sure of exactly what you mean here. This change _will_ disallow code like:

```ts
const getMagicNumber = () => 1234;
jest.beforeAll(() => getMagicNumber());
```

because we would then be returning a number. This is not void, undefined or Promise<unknown>.

This is intentional for Jest 27 to match @SimenB's change: ["Disallow return values other than a `Promise` from hooks and tests"](https://github.com/facebook/jest/pull/10512/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR23)

If you think we should allow hooks to return `any` we could easily implement that with something like:

```ts
type ProvidesHookCallback = (cb: DoneCallback) => any;
// or just so it's clear that it's a subtype (but wouldn't change the actual typing)
type ProvidesHookCallback = (cb: DoneCallback) => any | ProvidesCallback;

type Lifecycle = (fn: ProvidesHookCallback, timeout?: number) => any;
```

...but I believe this will create a discrepancy between DefinitelyTyped and Jest which is undesirable. Happy to take the consensus of active reviewers though.

Reply:
What I'm meaning is that it's still valid to return anything from a hook so long as you don't have `done` - I've confirmed this with this on latest jest:

```
describe('a test', () => {
  beforeEach(() => 1);

  it('passes', () => {
    expect(true).toBe(true);
  });
});
```

We're even doing it in [`eslint-plugin-jest`](https://github.com/jest-community/eslint-plugin-jest/blob/main/src/rules/__tests__/no-deprecated-functions.test.ts#L129):

```
// no-deprecated-functions.ts
/** @internal */
export const _clearCachedJestVersion = () => (cachedJestVersion = null);

// __tests_/no-deprecated-functions.ts
describe('the rule', () => {
  beforeEach(() => _clearCachedJestVersion());
  ...
});
```

The problem I'm seeing is that people will be using this pattern without issue on Jest 27, only for them to update their types and suddenly get it flagged as an error despite it working at runtime - and unless they scour the changelog and find the entry mentioning this breaking change, I expect they'll open issues and PRs here saying the types are wrong.

Reply:
Sure, happy to - I only made this change to match the changelog and internal jest types.

As long as nobody else objects I'll change this, as per the end of my previous comment (adding `ProvidesHookCallback`).

Reply:
> as per the end of my previous comment (adding ProvidesHookCallback).

I believe the change you described there won't be correct: `Lifecycle` is currently used to type both hooks (which do allow returning anything unless you're using `done`) & tests (which don't allow returning anything ever).

The current change you've made looks correct for tests, so I think we just need to make a new type (or really duplicate the types as they were before your change) for hooks - which is why I suggested maybe using a generic might be tidier since it would save having two `Lifecycle` types.

Reply:
What about when using `jest-jasmine2` as the test runner? In such cases, returning a `Promise` together with the `done` callback is completely ok and safe. I had to revert the version back exactly because of this.

We should rely on the runner reporting runtime errors instead of the type system, especially in cases where the runtime is configurable (like for jest). When using `jest-circus`, the combination of a done callback and a returned promise will raise a descriptive error:
```
Test functions cannot both take a 'done' callback and return something. Either use a 'done' callback, or return a promise.
```

```suggestion
    type ProvidesCallback = (cb: DoneCallback) => Promise<void | undefined> | void | undefined;
```

Reply:
Hi @enisdenjo, can you give an example of a test in `jest-jasmine2` that requires both a done callback and returning a promise?

As per the Jest [changelog](https://github.com/facebook/jest/blob/master/CHANGELOG.md) and [blog](https://jestjs.io/blog/2021/05/25/jest-27), breaking changes in Jest 27 include making TypeScript types stricter to prevent returning a Promise and taking a done callback.

Reply:
Hey @domdomegg,

> can you give an example of a test in jest-jasmine2 that requires both a done callback and returning a promise?

I use done+promise all around, here are a few examples:

- https://github.com/enisdenjo/graphql-ws/blob/06e3326e0467c91535a3d0697f249c1b7db8bfbd/src/__tests__/client.ts#L171-L192
- https://github.com/enisdenjo/graphql-ws/blob/06e3326e0467c91535a3d0697f249c1b7db8bfbd/src/__tests__/client.ts#L380-L402
- https://github.com/enisdenjo/graphql-ws/blob/06e3326e0467c91535a3d0697f249c1b7db8bfbd/src/__tests__/server.ts#L26-L59
- https://github.com/enisdenjo/graphql-ws/blob/06e3326e0467c91535a3d0697f249c1b7db8bfbd/src/__tests__/server.ts#L509-L532

Do note that all above examples [run on Jest v27](https://github.com/enisdenjo/graphql-ws/blob/06e3326e0467c91535a3d0697f249c1b7db8bfbd/package.json#L104) with the [test runner set to `jest-jasmine2`](https://github.com/enisdenjo/graphql-ws/blob/06e3326e0467c91535a3d0697f249c1b7db8bfbd/jest.config.js#L3).

This PR makes all of them fail with TS, even though its perfectly type legal and safe to do so with `jest-jasmine2`.

> As per the Jest changelog and blog, breaking changes in Jest 27 include making TypeScript types stricter to prevent returning a Promise and taking a done callback - no matter the runtime environment.

I think you're wrong here, I guess you're talking about this entry in the changelog:

>[jest-circus] [**BREAKING**] Fail tests when multiple `done()` calls are made (https://github.com/facebook/jest/pull/10624)

The change is exclusive to `jest-circus`, not whole `jest` (as stated in the scope on the beginning of the changelog entry). [`jest-circus` 
is a test runner](https://github.com/facebook/jest/tree/master/packages/jest-circus) and https://github.com/facebook/jest/pull/10624 fails tests during runtime.

> It might be worth raising an issue on Jest directly rather than the types if you disagree with the change to disallow using both a done callback and returning a promise.

Personally, I am against this change, it just complicates stuff unnecessarily. You want async tests to leverage the `await` and still use a `done` in a callback to indicate completion where tests have mixes of promises and events (as my examples show above).

Furthermore, there is already an issue raised for `jest-circus`: https://github.com/facebook/jest/issues/10529. Yeah, sure, you can just do https://github.com/facebook/jest/issues/10529#issuecomment-904608475; but, test suites are meant to be convenient and easy to use.

Reply:
My bad, I referenced the wrong changelog entry. You were probably talking about:

> [jest-globals] [BREAKING] Disallow mixing a `done` callback and returning a `Promise` from hooks and tests (https://github.com/facebook/jest/pull/10512)

which does indeed belong to `jest-globals`. 🤔

Reply:
@domdomegg even though the v27 changelog claims done+promise is disallowed globally, it seems not to be. The done+promise errors is only present when using the `jest-circus` test runner and the assertion can be found here (inside `jest-circus`): 

https://github.com/facebook/jest/blob/7c6cea2696e6f08553189a5e077126d610751576/packages/jest-circus/src/utils.ts#L212-L220